keps/sig-multicluster/2149-clusterid/README.md

```diff
@@ -0,0 +1,705 @@
+
+While a member of a ClusterSet, a cluster will also have an additional `clusterset.k8s.io ClusterClaim` which describes its current membership. This claim must be present exactly as long as the set membership lasts, and removed when the cluster is no longer a member.
+
+### User Stories
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### ClusterSet membership
+
+I have some set of clusters working together and need a way to uniquely identify them within the system that I use to track membership.
+
+#### Multi-Cluster Services
+
+I have a headless multi-cluster service deployed across clusters in my ClusterSet with similarly named pods in each cluster. I need a way to disambiguate each backend pod via DNS.
+
+#### Diagnostics
+
+Clusters within my ClusterSet send logs/metrics to a common monitoring solution and I need to be able to identify the cluster from which a given set of events originated.
+
+
+#### Joining or moving between ClusterSets
+
+I want the ability to add a previously-isolated cluster to a ClusterSet, or to move a cluster from one ClusterSet to another and be aware of this change.
+
+### `ClusterClaim` CRD
+
+The `ClusterClaim` resource provides a way to store identification related, cluster scoped information for multi-cluster tools while creating flexibility for implementations. A cluster may have multiple `ClusterClaim`s, each holding a different identification related value. Each claim contains the following information:
+
+*   **Name** - a well known or custom name to identify the claim.
+*   **Value** - a claim-dependent string.
+
+
+### Well known claims
+
+The requirements below use the keywords **must, should,** and **may** purposefully in accordance with [RFC-2119](https://tools.ietf.org/html/rfc2119).
+
+
+#### Claim: `id.k8s.io`
+
+Contains a unique identifier for the containing cluster.
+
+
+##### Uniqueness
+
+*   The identifier **must** be unique within the ClusterSet to which its cluster belongs for the duration of the cluster’s membership.
+*   The identifier **should** be unique beyond the ClusterSet within the scope of expected use.
+*   The identifier **may** be globally unique beyond the scope of its ClusterSet.
+*   The identifier **may** be unique beyond the span of its cluster’s membership and lifetime.
+
+
+##### Lifespan
+
+*   The identifier **must** exist and be immutable for the duration of a cluster’s membership in a ClusterSet, and as long as a `clusterset.k8s.io` claim exists.
+*   The identifier **must** exist for the lifespan of a cluster.
```

Comment:
This might be tricky to implement with the right value and also treat as immutable. Is it really linked this tightly with the lifecycle of the cluster itself, or is it more important in the context of the lifecycle of the cluster within the clusterset?

Reply:
I think any immutability requirements can only be scoped to a clusterset. I think the value of an ID drops significantly if we don't have at least that, but I don't think we can require that the ID is unique beyond the clusterset unless we make an even stronger statement - like the ID is some pre-existing well known UID (kube-system.uid).

I'm not sure we need to require the ID to exist beyond the clusterset membership either. We could probably use some example scenarios describing how some dependent tool should handle IDs changing or disappearing during various stages of the cluster/membership lifecycle

Reply:
Cleaned up the language and noted Jeremy's example scenarios request as unresolved in 75cab4d9e1bdc20f9bbd1cd086041d4a96103abb